### ISSUE TITLE
this test will failed, maybe is about footnote

### ISSUE THREAD

**Issue by xp44mm** [lunet-io/markdig#325]
the test is follow:
```F#
type MarkdigTest(output: ITestOutputHelper) =
    let text =
        Path.Combine(@"C:\Program Files\Typora\resources\app\Docs","Markdown Reference.md")
        |> File.ReadAllText

    let document =
        Markdown.Parse(text, MarkdownPipelineBuilder().UseAdvancedExtensions().Build())

    [<Fact>]
    member this.``NormalizeRenderer Test``() =
        use writer = new StringWriter()
        let normalizer = new NormalizeRenderer(writer)
        normalizer.Render(document) |> ignore
        let raw = writer.ToString()
        output.WriteLine(raw)

```
the test file is, need remove the .txt extent name:
[Markdown Reference.md.txt](https://github.com/lunet-io/markdig/files/3011949/Markdown.Reference.md.txt)

please note that `footnote`

**Comment by xoofx** [comment]
Can you explain what fails exactly?

**Comment by xp44mm** [comment]
![result](https://user-images.githubusercontent.com/10087796/55062114-3d095a00-50b0-11e9-8e83-150fd63da7a8.gif)

**Comment by xp44mm** [comment]
i remove some config:
```F#
    let document =
        let pipeline =
            MarkdownPipelineBuilder()//.UseAdvancedExtensions()
                .UseAbbreviations()
                //.UseAutoIdentifiers()//*
                .UseCitations()
                .UseCustomContainers()
                .UseDefinitionLists()
                .UseEmphasisExtras()
                .UseFigures()
                .UseFooters()
                //.UseFootnotes()//*
                .UseGridTables()
                .UseMathematics()
                .UseMediaLinks()
                .UsePipeTables()
                .UseListExtras()
                .UseTaskLists()
                .UseDiagrams()
                //.UseAutoLinks()//*
                .UseGenericAttributes()
                .Build()
        Markdown.Parse(text, pipeline)
```
maybe the **link reference definition** moves ome  to first occurency together from article bottom.

**Comment by xoofx** [comment]
Oh, yes, the `NormalizeRenderer` is by far not compatible with all extensions. It requires still lots of work and I have personally zero spare time left for that. so PR welcome. See also issue #155

**Comment by xp44mm** [comment]
left side is normed, right side is source.
![result](https://user-images.githubusercontent.com/10087796/55063927-02092580-50b4-11e9-901c-9deefa684138.gif)

**Comment by xoofx** [comment]
as I said, it is a known issue, look at #155 and you will see that most extensions are not supported by normalize, and even the core part of CommonMark might not be entirely supported well.

**Comment by xp44mm** [comment]
hello! i use regex and f# , C# i just read can't write. i read the code, i guess the bug not in normalize, insteed, bug in parser. 

maybe a `LinkReferenceDefinitionGroup` per every section. next code seem expression one group in a document. that will merged all groups into first. no respect source. 
https://github.com/lunet-io/markdig/blob/354db6b306559b73825140b33c3634137be062b1/src/Markdig/Syntax/LinkReferenceDefinitionExtensions.cs#L48

**Comment by MihaZupan** [comment]
Yes, the approach with a single `LinkReferenceDefinitionGroup` would have to be changed for the `NormalizeRenderer`, for the `HtmlRenderer` it was okay since the location of the definition does not matter for the overall document.

**Comment by xoofx** [comment]
> hello! i read the code, i guess the bug not in normalize, insteed, bug in parser. i use regex and f# , C# i just read can't write. manapulate char and use ref to modify source code beyond my job.

Yeah, as @MihaZupan suggested, originally markdig was mainly and only developed for HTML rendering. NormalizeRenderer has been added later, but is far from completed, and it will probably never be completed if nobody in the markdig userbase is willing to contribute fixing it. And that's lots of work.

**Comment by xp44mm** [comment]
yes, the NormalizeRenderer  is inverse of parser. it can help to test the parser if respect raw text. so holp to develop continue.
